server/src/com/cloud/network/router/NetworkHelperImpl.java

```diff
@@ -388,6 +388,19 @@ public DomainRouterVO startVirtualRouter(final DomainRouterVO router, final User
                     routerToBeAvoid = rrouter;
                 }
             }
+        } else if (router.getVpcId() != null) {
+            final List<DomainRouterVO> routerList = _routerDao.listByVpcId(router.getVpcId());
+            for (final DomainRouterVO rrouter : routerList) {
+                if (rrouter.getHostId() != null && rrouter.getIsRedundantRouter() && rrouter.getState() == State.Running) {
+                    if (routerToBeAvoid != null) {
+                        throw new ResourceUnavailableException("Try to start router " + router.getInstanceName() + "(" + router.getId() + ")"
+                                + ", but there are already two redundant routers with IP " + router.getPublicIpAddress() + ", they are " + rrouter.getInstanceName() + "("
+                                + rrouter.getId() + ") and " + routerToBeAvoid.getInstanceName() + "(" + routerToBeAvoid.getId() + ")", DataCenter.class,
+                                rrouter.getDataCenterId());
+                    }
+                    routerToBeAvoid = rrouter;
+                }
+            }
```

Comment:
nit - This could use clean up to avoid duplication of logic and fixing at multiple places -
```
       DomainRouterVO routerToBeAvoid = null;
       List<DomainRouterVO> routerList = ;
        if (networkIds.size() != 0) {
            routerList = _routerDao.findByNetwork(networkIds.get(0));
        } else if (router.getVpcId() != null) {
            routerList = _routerDao.listByVpcId(router.getVpcId());
        }
        if (CollectionUtils.isNotEmpty(routerList) {
           for (final DomainRouterVO rrouter : routerList) {
                if (rrouter.getHostId() != null && rrouter.getIsRedundantRouter() && rrouter.getState() == State.Running) {
                    if (routerToBeAvoid != null) {
                        throw new ResourceUnavailableException("Try to start router " + router.getInstanceName() + "(" + router.getId() + ")"
                                + ", but there are already two redundant routers with IP " + router.getPublicIpAddress() + ", they are " + rrouter.getInstanceName() + "("
                                + rrouter.getId() + ") and " + routerToBeAvoid.getInstanceName() + "(" + routerToBeAvoid.getId() + ")", DataCenter.class,
                                rrouter.getDataCenterId());
                    }
                    routerToBeAvoid = rrouter;
                }
           }
     }
```

Reply:
updated code as @anuragaw suggested